### PR TITLE
Fix http.response.template being used with BaseHTTPMiddleware

### LINKED ISSUE #472: AssertionError with middleware and TemplateResponse
```diff
diff --git i/tests/test_templates.py w/tests/test_templates.py
index b48373e..57dee91 100644
--- i/tests/test_templates.py
+++ w/tests/test_templates.py
@@ -20,6 +20,11 @@ def test_templates(tmpdir):
     async def homepage(request):
         return templates.TemplateResponse("index.html", {"request": request})

+    @app.middleware("http")
+    async def noop(request, call_next):
+        response = await call_next(request)
+        return response
+
     client = TestClient(app)
     response = client.get("/")
     assert response.text == "<html>Hello, <a href='http://testserver/'>world</a></html>"
```

causes:
```
tests/test_templates.py:29: in test_templates
    response = client.get("/")
.venv/lib/python3.7/site-packages/requests/sessions.py:546: in get
    return self.request('GET', url, **kwargs)
starlette/testclient.py:421: in request
    json=json,
.venv/lib/python3.7/site-packages/requests/sessions.py:533: in request
    resp = self.send(prep, **send_kwargs)
.venv/lib/python3.7/site-packages/requests/sessions.py:646: in send
    r = adapter.send(request, **kwargs)
starlette/testclient.py:238: in send
    raise exc from None
starlette/testclient.py:235: in send
    loop.run_until_complete(self.app(scope, receive, send))
/usr/lib/python3.7/asyncio/base_events.py:584: in run_until_complete
    return future.result()
starlette/applications.py:134: in __call__
    await self.error_middleware(scope, receive, send)
starlette/middleware/errors.py:122: in __call__
    raise exc from None
starlette/middleware/errors.py:100: in __call__
    await self.app(scope, receive, _send)
starlette/middleware/base.py:25: in __call__
    response = await self.dispatch_func(request, self.call_next)
tests/test_templates.py:21: in noop
    response = await call_next(request)
starlette/middleware/base.py:47: in call_next
    assert message["type"] == "http.response.start"
E   AssertionError
```
`message` is `{'type': 'http.response.template', 'template': <Template 'index.html'>, 'context': {'request': <starlette.requests.Request object at 0x7ff4a2d21e48>}}` there.

This appears to be an issue with the test client only though.
Do I understand it correctly that the "http.response.template" extension is only used with tests?

### PR DESCRIPTION
**Pull request by blueyed**
Fixes #472.

### CODE REVIEW

**Comment by blueyed** [comment]
@tomchristie 
Would appreciate some insight here.

File: tests/test_templates.py [hunk 1]
```diff
@@ -3,12 +3,12 @@
 import pytest
 
 from starlette.applications import Starlette
-from starlette.responses import HTMLResponse
 from starlette.templating import Jinja2Templates
 from starlette.testclient import TestClient
 
 
-def test_templates(tmpdir):
+@pytest.mark.parametrize("with_middleware", ("with", "without"))
```
Comment (lovelydinosaur) [inline tests/test_templates.py]: Let's just create a seperate test case.

**Comment by blueyed** [comment]
Only style issues.

Awaiting feedback before fixing this.

It is not very helpful btw ;) (should show the diff)

> +black starlette tests setup.py --check
would reformat /home/travis/build/encode/starlette/starlette/middleware/base.py
All done! 💥 💔 💥
1 file would be reformatted, 60 files would be left unchanged.

**Comment by em92** [comment]
> It is not very helpful btw ;) (should show the diff)

Running `./scripts/lint` does reformat code

File: starlette/middleware/base.py [hunk 2]
```diff
@@ -44,14 +46,20 @@ def __init__(self, app: ASGIApp, dispatch: DispatchFunction = None) -> None:
         if message is None:
             task.result()
             raise RuntimeError("No response returned.")
-        assert message["type"] == "http.response.start"
+
+        if "http.response.template" in scope.get("extensions", {}):
```
Comment (blueyed) [inline starlette/middleware/base.py:50]: This is always True (https://codecov.io/gh/blueyed/starlette/compare/ef49e73e1043fdf7ce409a190ab632cfd568953e...e3ec97fcb16de9f3bb645ddb67495349c124dbb2/diff#D4-50).
I.e. we should maybe test with a TestClient that does not use it?!

**Comment by lovelydinosaur** [comment]
On reflection, the smarter thing for us to do in middleware would be to *wait* for the first body message, before figuring out if we should treat it as a streaming response, or a standard response.
If we did that, then we could use `Response`, `TemplateResponse`, or `StreamingResponse` depending on the messages coming down the channel.

**Pull request closed without merging**